pandas/io/excel/_openpyxl.py

```diff
@@ -451,3 +462,277 @@ def write_cells(self, cells, sheet_name=None, startrow=0, startcol=0,
+                if len(index_col) == 1:
+                    # TODO: understand why this is needed
+                    raise TypeError("list indices must be integers.*, not str")
+            else:
+                frame = frame.set_index([column_names[i] for i in index_col])
+        else:
+            if isinstance(index_col, str):
+                frame = frame.set_index(index_col)
+            else:
+                frame = frame.set_index(column_names[index_col])
+        return frame
+
+    def get_sheet_by_name(self, name):
+        return self.book[name]
+
+    def get_sheet_by_index(self, index):
+        return self.book.worksheets[index]
+
+    def _replace_type_error_with_nan(self, rows):
+        try:
+            from openpyxl.cell.cell import TYPE_ERROR
+        except ImportError:
+            # Work around for import error in Linux py35_compat test that I
+            # can't reproduce
+            TYPE_ERROR = 'e'
+
+        nan = float('nan')
+        for row in rows:
+            yield [nan
+                   if cell.data_type == TYPE_ERROR
+                   else cell.value
+                   for cell in row]
+
+    def get_sheet_data(self, sheet):
+        data = self._replace_type_error_with_nan(sheet.rows)
+        return list(data)
+
+    def _parse_sheet(self, sheet, convert_float, usecols, header, skiprows,
+                     index_col, converters, skipfooter, dtype, squeeze):
+        """Parse a single sheet into a dataframe."""
+
+        data = self.get_sheet_data(sheet)
+        if not data or data == [[None]]:
+            return DataFrame()
+
+        usecols = _maybe_convert_usecols(usecols)
+
+        if is_list_like(header) and len(header) == 1:
+            header = header[0]
+
+        header_names = self._handle_header_keywords(data, header, skiprows,
+                                                    index_col)
+
+        # TODO: implement whatever this should do
+        # has_index_names = is_list_like(header) and len(header) > 1
+
+        if skiprows:
+            data = [row for i, row in enumerate(data) if i not in skiprows]
+
+        if skipfooter:
+            data = data[:-skipfooter]
+
+        column_names = [cell for cell in data.pop(0)]
+
+        frame = DataFrame(data, columns=column_names)
+        frame = self._handle_usecols(frame, usecols)
+
+        if not converters:
+            converters = dict()
+        if not dtype:
+            dtype = dict()
+
+        # handle columns referenced by number so all references are by
+        # column name
+        handled_converters = {}
+        for k, v in converters.items():
+            if k not in frame.columns and isinstance(k, int):
+                k = frame.columns[k]
+            handled_converters[k] = v
+        converters = handled_converters
+
+        if len(frame) > 0:
+            for column in set(frame) - set(dtype.keys()):
+                frame[column] = self._handle_convert_float(frame[column],
```

Comment:
Rather than using a custom function is it possible to emulate the xlrd code for this, namely looking at the type of the cell? There are slight nuances to doing this in a vectorized approach vs a cell by cell (which is what xlrd does, for better or for worse) so I think we should keep these consistent for now.

Refactor of this would probably get around the CI failure you are currently hitting